### PR TITLE
FINERACT-2081: introduce builder pattern to Note entity instantiation.

### PR DESCRIPTION
**Pull request by Zeyad2003**
## Description

- Removed all redundant static instantiation methods and constructors from `Note.java`.
- Used Builder pattern to make it easy to instantiate the Note object.
- Changed all instantiation places to use the new implementation.

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java [hunk 1]
```diff
@@ -37,150 +38,46 @@
 import org.apache.fineract.portfolio.shareaccounts.domain.ShareAccount;
 
 @Entity
+@Getter
+@Builder
+@NoArgsConstructor
```
Comment (kjozsa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java]: do we need the public no-arg and all-args constructors if we plan to switch to Builder?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/note/service/NoteWritePlatformServiceJpaRepositoryImpl.java [hunk 2]
```diff
@@ -456,4 +472,14 @@ private Note getNoteForDelete(final JsonCommand command) {
         return noteForUpdate;
     }
 
+    public Map<String, Object> update(final JsonCommand command, Note noteForUpdate) {
+        final Map<String, Object> actualChanges = new LinkedHashMap<>(7);
+        final String newNote = command.stringValueOfParameterNamed("note");
+        if (!noteForUpdate.getNote().equals(newNote)) {
```
Comment (kjozsa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/note/service/NoteWritePlatformServiceJpaRepositoryImpl.java:478]: are you sure the equals() will catch any content changes inside a Note? We might need a unit test for this logic as well.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java (hunk 1 above)
Comment (Zeyad2003) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java]: Yes, the `@Builder` annotation requires a proper constructor for initializing the entity>

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/note/service/NoteWritePlatformServiceJpaRepositoryImpl.java (hunk 2 above)
Comment (Zeyad2003) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/note/service/NoteWritePlatformServiceJpaRepositoryImpl.java:478]: This is the same old logic, just moved from the entity to the service layer for better separation of concerns. 

The `equals()` method will work as will correctly identify changes in the note content, as it compares the entire string.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java (hunk 1 above)
Comment (kjozsa) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java]: my point is that we should not have public constructors, if we are switching to the builder pattern. Please make sure that constructors are having the proper access level.

Comment (Zeyad2003) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/Note.java]: Done.

**Comment by Zeyad2003** [comment]
@kjozsa

Are there any required updates?

**Comment by kjozsa** [comment]
@Zeyad2003 I'm fine, but I recall @vidakovic also had concerns with the nature of the changes. Let's get him review your work as well before we proceed here.

**Pull request closed without merging**